src/catalog/query_history_catalog.cpp

```diff
@@ -36,22 +34,23 @@ QueryHistoryCatalog::QueryHistoryCatalog(concurrency::TransactionContext *txn)
                       "timestamp      TIMESTAMP NOT NULL);",
                       txn) {}
 
-QueryHistoryCatalog::~QueryHistoryCatalog() {}
+QueryHistoryCatalog::~QueryHistoryCatalog() = default;
 
-bool QueryHistoryCatalog::InsertQueryHistory(const std::string &query_string, 
-                                  std::string &fingerprint, uint64_t timestamp,
-                                  type::AbstractPool *pool,
-                                  concurrency::TransactionContext *txn) {
+bool QueryHistoryCatalog::InsertQueryHistory(
+    const std::string &query_string, const std::string &fingerprint,
+    uint64_t timestamp, type::AbstractPool *pool,
+    concurrency::TransactionContext *txn) {
   std::unique_ptr<storage::Tuple> tuple(
       new storage::Tuple(catalog_table_->GetSchema(), true));
 
-  auto val0 = type::ValueFactory::GetVarcharValue(query_string, pool);
-  auto val1 = type::ValueFactory::GetVarcharValue(fingerprint, pool);
+  auto val0 = type::ValueFactory::GetVarcharValue(query_string);
+  auto val1 = type::ValueFactory::GetVarcharValue(fingerprint);
   auto val2 = type::ValueFactory::GetTimestampValue(timestamp);
 
-  tuple->SetValue(ColumnId::QUERY_STRING, val0, pool);
-  tuple->SetValue(ColumnId::FINGERPRINT, val1, pool);
-  tuple->SetValue(ColumnId::TIMESTAMP, val2, pool);
+  tuple->SetValue(ColumnId::QUERY_STRING, val0,
+                  pool != nullptr ? pool : &pool_);
+  tuple->SetValue(ColumnId::FINGERPRINT, val1, pool != nullptr ? pool : &pool_);
+  tuple->SetValue(ColumnId::TIMESTAMP, val2, pool != nullptr ? pool : &pool_);
```

Comment:
Actually could you remind me what's the difference between using the `pool` pointer from the argument, using `pool_` from the object itself, and just passing in a `nullptr`?

Reply:
A pool has to be used, either the one from the function argument or the member pool, otherwise memory is `new`-ed inside `SetValue` that is never freed.

Reply:
I see! And when to use the one from the function argument, and when the member pool?

Reply:
There is a much deeper rooted problem w.r.t. string handling. All these pools never claim memory until they're destroyed, meaning they don't reuse memory. Someone needs to fix this ...

Reply:
Okay...

Reply:
Thanks for the clarification!